vm/src/stdlib/itertools.rs

```diff
@@ -181,6 +184,141 @@ impl PyItertoolsTakewhile {
+                        return Err(vm.new_value_error("Stop argument for islice() must be None or an integer: 0 <= x <= sys.maxsize.".to_string()));
+                    }
+                    stop
+                } else {
+                    None
+                };
+
+                (iter, 0usize, stop, 1usize)
+            }
+            _ => {
+                let (iter, start, stop, step): (
+                    PyObjectRef,
+                    PyObjectRef,
+                    PyObjectRef,
+                    PyObjectRef,
+                ) = args.bind(vm)?;
+                let stop = if !stop.is(&vm.get_none()) {
+                    let stop = pyobject_to_usize(stop, &vm);
+                    if !stop.is_some() {
+                        return Err(vm.new_value_error("Stop argument for islice() must be None or an integer: 0 <= x <= sys.maxsize.".to_string()));
+                    }
+                    stop
+                } else {
+                    None
+                };
+
+                let start = if !start.is(&vm.get_none()) {
+                    let start = pyobject_to_usize(start, &vm);
+                    if !start.is_some() {
+                        return Err(vm.new_value_error("Indices for islice() must be None or an integer: 0 <= x <= sys.maxsize.".to_string()));
+                    }
+                    start.unwrap()
+                } else {
+                    0usize
+                };
+
+                let step = if !step.is(&vm.get_none()) {
+                    let step = pyobject_to_usize(step, &vm);
+                    if !step.is_some() {
+                        return Err(vm.new_value_error(
+                            "Step for islice() must be a positive integer or None.".to_string(),
+                        ));
+                    }
+                    step.unwrap()
+                } else {
+                    1usize
+                };
+
+                (iter, start, stop, step)
+            }
+        };
+
+        let iter = get_iter(vm, &iter)?;
+
+        Ok(PyItertoolsIslice {
+            iterable: iter,
+            cur: RefCell::new(0),
+            next: RefCell::new(start),
+            stop: stop,
+            step: step,
+        }
+        .into_ref(vm)
+        .into_object())
+    }
+
+    #[pymethod(name = "__next__")]
+    fn next(&self, vm: &VirtualMachine) -> PyResult {
+        while *self.cur.borrow() < *self.next.borrow() {
+            call_next(vm, &self.iterable)?;
+            *self.cur.borrow_mut() += 1;
+        }
+
+        if self.stop.is_some() && *self.cur.borrow() >= self.stop.unwrap() {
+            return Err(new_stop_iteration(vm));
+        }
+
+        let obj = call_next(vm, &self.iterable)?;
+        *self.cur.borrow_mut() += 1;
+
+        // TODO is this overflow check required? attempts to copy CPython but it's
+        // currently broken.
+        let (next, ovf) = (*self.next.borrow()).overflowing_add(self.step);
```

Comment:
You can call `borrow_mut` once for this bit, then reuse the borrow.
How about
```rust
let next = self.next.borrow_mut();
*next = *next.checked_add(self.step).or(self.stop).ok_or_else(|| new_vm_error())?;
```

But why are you checking for overflow here, if you didn't check for overflow in the `+= 1` lines?

Reply:
Basically followed CPython's `itertools` on this, which doesn't check for an overflow here.

Technically CPython's `itertools` is wrong because after an overflow on `next` when `stop=None`, `next` is reset to `-1` which is incorrect because `cnt` itself will overflow successfully to `ssize_t`'s min (If it won't be optimized away due to its UB being), and so the next object return by `islice` will be after a random step.

Now all of this is purely hypothetical here since we use `usize` which will usually take the form of 64 bits.